src/ReinforcementLearningZoo/src/algorithms/policy_gradient/trpo.jl

```diff
@@ -0,0 +1,113 @@
+export TRPO
+
+using Random: GLOBAL_RNG, shuffle, AbstractRNG
+using Functors: @functor
+using Flux: Flux, destructure
+using StatsBase: mean
+using ChainRulesCore: ignore_derivatives
+
+"""
+Trust Region Policy Optimization
+"""
+Base.@kwdef struct TRPO{A,B,D} <: AbstractPolicy
+    approximator::A
+    baseline::B
+    "only a discrete action space is supported for now"
+    dist::D = Distributions.Categorical
+    γ::Float32 = 0.99f0
+    batch_size::Int = 1024
+    rng::AbstractRNG = GLOBAL_RNG
+    max_backtrack_step::Int = 10
+    kldivergence_limit::Float32 = 1f-2
+    backtrack_coeff::Float32 = 0.8f0
+    cg_max_iter::Int = 10
+    cg_ϵ::Float32 = 1f-10
+    damping_coeff = 1f-2
+end
+
+IsPolicyGradient(::Type{<:TRPO}) = IsPolicyGradient()
+@functor TRPO (approximator, baseline)
+
+function (π::TRPO)(env::AbstractEnv)
+    res = env |> state |> send_to_device(π) |> π.approximator |> send_to_host
+    rand(π.rng, action_distribution(π.dist, res)[1])
+end
+
+function (p::Agent{<:TRPO})(::PostEpisodeStage, env::AbstractEnv)
+    p.trajectory.container[] = true
+    optimise!(p.policy, p.trajectory.container)
+    empty!(p.trajectory.container)
+end
+
+RLBase.optimise!(::Agent{<:TRPO}) = nothing
+
+function RLBase.optimise!(π::TRPO, episode::Episode)
+    gain = discount_rewards(episode[:reward][:], π.γ)
+    for inds in Iterators.partition(shuffle(π.rng, 1:length(episode)), π.batch_size)
+        optimise!(π, (state=episode[:state][inds], action=episode[:action][inds], gain=gain[inds]))
+    end
+end
+
+function RLBase.optimise!(p::TRPO, batch::NamedTuple{(:state, :action, :gain)})
+    A = p.approximator
+    B = p.baseline
+    s, a, g = map(Array, batch) # !!! FIXME
+
+    # fit value network on mean-square
+    # store δ for advantage estimate
+    if isnothing(B)
+        δ = normalise(g)
+    else
+        gs = gradient(params(B)) do
+            δ = g - vec(B(s))
+            loss = mean(δ .^ 2)
+            ignore_derivatives() do
+                # @info "TRPO/baseline" loss = loss δ
+            end
+            loss
+        end
+        optimise!(B, gs)
+    end
+    
+    # store logits as intermediate value
+    old_logits = Ref{Matrix{Float32}}()
```

Comment:
there were some oddities with Zygote 2nd order derivatives wrt implicit parameters when i tried `local old_logits`, yielding inconsistent results after `mapreduce(vec, vcat, gradient)` (i got a `Ref(0)` term sometimes as the first term).  i'm not sure this is necessary anymore however, since i've since changed how the 2nd order gradient is calculated.